snap/hooks/configure

```diff
@@ -1,158 +1,90 @@
-    echo ""                                                                   >> "${kiosk_display}~"
-    echo "${display_config}"                                                  >> "${kiosk_display}~"
-    if [ -e "${kiosk_display}" ]; then
-      if ! diff "${kiosk_display}~" "${kiosk_display}" > /dev/null; then
-        mv "${kiosk_display}"  "${kiosk_display}.save"
-        mv "${kiosk_display}~" "${kiosk_display}"
-        let display_changes+=1
-      else
-        rm "${kiosk_display}~"
-      fi
-    else
-      mv "${kiosk_display}~" "${kiosk_display}"
-      let display_changes+=1
-    fi
-  else
-    if [ -e "${kiosk_display}" ]; then
-      if grep "# USE 'snap set mir-kiosk display-config=...' INSTEAD" "${kiosk_display}"; then
-        mv "${kiosk_display}"  "${kiosk_display}.save"
-        let display_changes+=1
-      fi
-    fi
-  fi
-fi
-
-# display-layout
-display_layout=$(snapctl get display-layout)
-if [ -n "${display_layout}" ]; then
-  if [ -e "${kiosk_display}" ]; then
-    if [ -z "$(grep "^  ${display_layout}:" "${kiosk_display}")" ]; then
-      echo "ERROR: '$display_layout' is not a layout in ${kiosk_display}:"
-      echo "---"
-      cat "${kiosk_display}"
-      echo "---"
-      exit 1
-    fi
-  else
-    if [ "${display_layout}" != "default" ]; then
-      echo "ERROR: '$display_layout' is not a layout in ${kiosk_display}:"
-      echo "---"
-      cat "${kiosk_display}"
-      echo "---"
-      exit 1
-    fi
-  fi
-
-  if [ "display-layout=${display_layout}" != "$(grep \^display-layout= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^display-layout=/d' ${kiosk_config}
-    else
-      sed --in-place=.save '/^display-layout=/d' ${kiosk_config}
-    fi
-    echo display-layout=${display_layout} >> ${kiosk_config}
+# $3
+$1=${2}
+EOT
     let config_changes+=1
   fi
-fi
+}
 
-# vt
-vt=$(snapctl get vt)
-if [ -n "${vt}" ]; then
-  if [ -n "${vt//[0-9]}" ] || [ ! -e "/dev/tty$vt" ]; then
-    echo "ERROR: '$vt' is not a valid VT"
-    exit 1
-  fi
+# display_file
+if kiosk_display=$(snapctl get kiosk-display); then
+  # A display configuration can be supplied through the kiosk-display option
+  if [ -n "${kiosk_display}" ]; then
+    cat <<EOT >> "${display_file}~"
+# DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
+# USE 'snap set $SNAP_NAME kiosk-display=...' INSTEAD
 
-  if [ "vt=${vt}" != "$(grep vt= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^vt/d' ${kiosk_config}
+${kiosk_display}
+EOT
+    if ! diff "${display_file}~" "${display_file}" > /dev/null; then
+      mv "${display_file}"  "${display_file}.save"
+      mv "${display_file}~" "${display_file}"
+      let config_changes+=1
     else
-      sed --in-place=.save '/^vt/d' ${kiosk_config}
+      rm "${display_file}~"
     fi
-    echo vt=${vt} >> ${kiosk_config}
-    let config_changes+=1
   fi
 fi
```

Comment:
You could have a common function taking the config key, file path and default contents to deal with both the display and config?